applications/FluidDynamicsApplication/custom_utilities/mass_conservation_utility.cpp

```diff
@@ -246,75 +317,97 @@ void MassConservationCheckProcess::ComputeVolumesAndInterface( double& positiveV
+                const double min_distance = r_geom[i].GetValue(NODAL_H)*1e-6;
+                if ( std::abs(r_dist) < min_distance ){
+                    r_geom[i].SetLock();
+                    r_dist = r_dist > 0 ? min_distance : -min_distance;
+                    r_geom[i].UnSetLock();
+                }
+                distance[i] = r_dist;
             }
 
-            if ( rGeom.PointsNumber() == 3 ){ p_modified_sh_func = Kratos::make_unique<Triangle2D3ModifiedShapeFunctions>(it_elem->pGetGeometry(), Distance); }
-            else if ( rGeom.PointsNumber() == 4 ){ p_modified_sh_func = Kratos::make_unique<Tetrahedra3D4ModifiedShapeFunctions>(it_elem->pGetGeometry(), Distance); }
-            else { KRATOS_ERROR << "The process can not be applied on this kind of element" << std::endl; }
+            const auto p_modified_sh_func = GetModifiedShapeFunctions(it_elem->pGetGeometry(), distance);
 
             // Concerning their area, the positive and negative side of the interface are equal
-            p_modified_sh_func->ComputeInterfacePositiveSideShapeFunctionsAndGradientsValues(
+            p_modified_sh_func->ComputeInterfaceNegativeSideShapeFunctionsAndGradientsValues(
                     shape_functions,                    // N
                     shape_derivatives,                  // DN
                     w_gauss_interface,                  // includes the weights of the GAUSS points (!!!)
-                    GeometryData::GI_GAUSS_1);          // first order Gauss integration
+                    GeometryData::GI_GAUSS_2);          // second order Gauss integration
+
+            // negative side outwards area normal vector values for the Gauss pts. of given quadrature
+            std::vector<Vector> normal_vectors;
+            p_modified_sh_func->ComputeNegativeSideInterfaceAreaNormals(
+                normal_vectors,
+                GeometryData::GI_GAUSS_2
+            );
+
+            // iteration over all 3 or 6 integration points
+            for ( unsigned int i_gauss = 0; i_gauss < w_gauss_interface.size(); i_gauss++)
+            {
+                const double& r_weight = w_gauss_interface[i_gauss];
+                const auto& r_N = row(shape_functions, i_gauss);
+
+                auto& r_normal = normal_vectors[i_gauss];
+                if (norm_2( r_normal ) > 1e-15) {
+                    r_normal /= norm_2( r_normal );
+                }
+                r_normal *= Factor;
 
-            for ( unsigned int i = 0; i < w_gauss_interface.size(); i++){
-                int_area += std::abs( w_gauss_interface[i] );
+                array_1d<double,3> interpolated_velocity = ZeroVector(3);
+                for (unsigned int n_node = 0; n_node < r_geom.PointsNumber(); n_node++){
+                    noalias( interpolated_velocity ) += r_N[n_node] * (r_geom[n_node].FastGetSolutionStepValue(VELOCITY)
+                        - r_geom[n_node].FastGetSolutionStepValue(MESH_VELOCITY));
+                }
```

Comment:
```suggestion
                for (unsigned int i_node = 0; i_node < r_geom.PointsNumber(); i_node++) {
                    const auto& r_node = r_geom[i_node];
                    noalias( interpolated_velocity ) += r_N[i_node] * r_node.FastGetSolutionStepValue(VELOCITY)                         - r_node.FastGetSolutionStepValue(MESH_VELOCITY);
                }
```
(`n_node` can be easily misunderstood as the total number of nodes).